src/rect.rs

```diff
@@ -455,9 +460,137 @@ impl<T: NumCast + Copy, Unit> TypedRect<T, Unit> {
     }
 }
 
+/// A 2d Rectangle represented using two points, optionally tagged with a unit.
+#[repr(C)]
+pub struct TypedBox2D<T, Unit = UnknownUnit> {
+    min: TypedPoint2D<T, Unit>,
+    max: TypedPoint2D<T, Unit>,
+}
+
+/// The default box type with no unit.
+pub type Box2D<T> = TypedBox2D<T, UnknownUnit>;
+
+impl<T, U> TypedBox2D<T, U> {
+    pub fn new(min: TypedPoint2D<T, U>, max: TypedPoint2D<T, U>) -> Self {
+        TypedBox2D {
+            min: min,
+            max: max,
+        }
+    }
+}
+
+impl<T, U> TypedBox2D<T, U>
+where T: Copy + Clone + Zero + One + PartialOrd + PartialEq + Add<T, Output=T> + Sub<T, Output=T> + Mul<Output=T> {
+    #[inline]
+    pub fn to_rect(&self) -> Option<TypedRect<T, U>> {
+        if self.is_empty_or_negative() {
+            return None;
+        }
+
+        Some(TypedRect {
+            origin: self.min,
+            size: self.size(),
+        })
+    }
+
+    #[inline]
+    pub fn size(&self) -> TypedSize2D<T, U> {
+        (self.max - self.min).to_size()
+    }
+
+    #[inline]
+    pub fn is_empty_or_negative(&self) -> bool {
+        let size = self.size();
+        let _0 = T::zero();
+
+        size.width <= _0 && size.height <= _0
+    }
+
+    #[inline]
+    pub fn intersects(&self, other: &Self) -> bool {
+        self.min.x < other.max.x && other.min.x < self.max.x &&
+            self.min.y < other.max.y && other.min.y <  self.max.y
+    }
+
+    #[inline]
+    pub fn intersection(&self, other: &Self) -> Self {
+        let min_point = point2(max(self.min.x, other.min.x), max(self.min.y, other.min.y));
+        let max_point = point2(min(self.max.x, other.max.x), min(self.max.y, other.max.y));
+
+        Self::new(min_point, max_point)
+    }
+
+    #[inline]
+    pub fn union(&self, other: &Self) -> Self {
+        if other.is_empty_or_negative() {
```

Comment:
I don't think we can rely on the maxes for union (although it does work for intersection). We could turn this into a `debug_assert` and force people to use ensure_positive beforehand if this is a performance concern.